gateway/blocks_gateway.go

```diff
@@ -223,38 +231,208 @@ func (api *BlocksGateway) Head(ctx context.Context, path ImmutablePath) (Content
 	return md, fileNode, nil
 }
 
-func (api *BlocksGateway) GetCAR(ctx context.Context, path ImmutablePath) (ContentPathMetadata, io.ReadCloser, <-chan error, error) {
-	// Same go-car settings as dag.export command
-	store := dagStore{api: api, ctx: ctx}
+func (api *BlocksGateway) GetCAR(ctx context.Context, p ImmutablePath, params CarParams) (io.ReadCloser, <-chan error, error) {
+	contentPathStr := p.String()
+	if contentPathStr[0:6] != "/ipfs/" {
+		return nil, nil, fmt.Errorf("path does not have /ipfs/ prefix")
+	}
+	firstSegment, _, _ := strings.Cut(contentPathStr[6:], "/")
+	rootCid, err := cid.Decode(firstSegment)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	// TODO: When switching to exposing path blocks we'll want to add these as well
-	roots, lastSeg, err := api.getPathRoots(ctx, path)
+	r, w := io.Pipe()
+	cw, err := storage.NewWritable(w, []cid.Cid{rootCid}, carv2.WriteAsCarV1(true))
 	if err != nil {
-		return ContentPathMetadata{}, nil, nil, err
+		return nil, nil, err
 	}
 
-	md := ContentPathMetadata{
-		PathSegmentRoots: roots,
-		LastSegment:      lastSeg,
+	blockGetter := merkledag.NewDAGService(api.blockService).Session(ctx)
+
+	blockGetter = &nodeGetterToCarExporer{
+		ng: blockGetter,
+		cw: cw,
 	}
 
-	rootCid := lastSeg.Cid()
+	// Setup the UnixFS resolver.
+	f := newNodeGetterFetcherSingleUseFactory(ctx, blockGetter)
+	pathResolver := resolver.NewBasicResolver(f)
 
-	// TODO: support selectors passed as request param: https://github.com/ipfs/kubo/issues/8769
-	// TODO: this is very slow if blocks are remote due to linear traversal. Do we need deterministic traversals here?
-	dag := car.Dag{Root: rootCid, Selector: selectorparse.CommonSelector_ExploreAllRecursively}
-	c := car.NewSelectiveCar(ctx, store, []car.Dag{dag}, car.TraverseLinksOnlyOnce())
-	r, w := io.Pipe()
+	lsys := cidlink.DefaultLinkSystem()
+	unixfsnode.AddUnixFSReificationToLinkSystem(&lsys)
 
 	errCh := make(chan error, 1)
 	go func() {
-		carWriteErr := c.Write(w)
+		// TODO: support selectors passed as request param: https://github.com/ipfs/kubo/issues/8769
+		// TODO: this is very slow if blocks are remote due to linear traversal. Do we need deterministic traversals here?
+		carWriteErr := walkGatewaySimpleSelector(ctx, ipfspath.Path(contentPathStr), params, &lsys, pathResolver)
```

Comment:
I was wondering exactly this while looking at the code; just to be clear—you're saying that the blocker here is the byte ranges? I haven't played with it myself yet (hasn't quite risen to our priority list in lassie) but it's supposed to be something we can do:

* https://github.com/ipld/ipld/pull/184
* https://github.com/ipld/go-ipld-prime/pull/375

There's an awful lot of code in here that's already handle by unixfsnode that would be good to offload. If there's things we can address to make this more usable then we should try.

Pinging @hannahhoward @willscott.

Here's a full form of the traverser and CAR generator that's used in some integration tests, for reference on how the whole thing can fit together, it doesn't have byte ranges though: https://github.com/filecoin-project/lassie/blob/b5aa6be20126af642eb7a86d96144e0f92dd2976/pkg/internal/itest/testpeer/generator.go#L366-L438

Reply:
Any update on this @rvagg @aschmahmann?

Reply:
Last discussion was agreement that we can't do enough of it with go-ipld-prime, that there are some awkward corners. I think more could be less manual than what's done here but the functionality gaps are enough to spur additional work. For now this implementation will have to do I suppose.
Currently @hannahhoward and I are just figuring out who's going to free up first enough to dive in to making this properly work through selectors, or some other combination of pieces that makes this more straightforward, because we'd like it for Lassie too, or at least the CAR verifier we have that we want to reuse in a bunch of places that these CARs flow through.